### ISSUE TITLE
[FEA] reduce retry window for Project

### ISSUE THREAD

**Issue by revans2** [NVIDIA/spark-rapids#7868]
**Is your feature request related to a problem? Please describe.**
Once https://github.com/NVIDIA/spark-rapids/pull/7864 and https://github.com/NVIDIA/spark-rapids/issues/7672 are done it would be very nice to have the ability to reduce what is retried when processing a project operation.  Right now if a project fails all of the expressions will be re-evaluated.  This is done because it is expensive to make the completed work spillable. But #7672 should solve that cost problem. 

The goal of this would be to make the result of each expressions spillable as soon as it completes and save it away so if we have to retry an operation we only have to retry what was not completed.  I don't know if this is actually going to be a performance win or not. We should do some performance tests once a prototype is ready because it might be more expensive to spill/unspill the result, than it would be to re-calculate it. I think on average it is likely to be a win, but we need to test this.

If it does look like it is not a win, then perhaps we can see if a heuristic would help. Something like doing a checkpoint if the computation rate is slower than the PCIe spill rate would be.
